File: packages/vscode-extension/src/pricingConfig.ts
import { join } from "node:path";
import { loadPricingTable, type ModelPricing, type PricingTable } from "@token-tracker/engine";

/** Kształt sekcji `tokenTracker.*` z ustawień VS Code — surowe, niezaufane dane użytkownika. */
export interface PricingConfig {
  pricingFile?: string;
  pricingOverrides?: unknown;
}

export interface ResolvedPricing {
  table: PricingTable;
  /** Klucze z `pricingOverrides`, które miały niepoprawny kształt i zostały pominięte. */
  rejectedModels: string[];
}

const PRICE_FIELDS: readonly (keyof ModelPricing)[] = [
  "inputPer1M",
  "outputPer1M",
  "cacheReadPer1M",
  "cacheWritePer1M",
];

export function resolvePricingTable(config: PricingConfig, homeDir: string): ResolvedPricing {
  const { valid, rejected } = parsePricingOverrides(config.pricingOverrides);
  return { table: { ...defaultPricingTable(config, homeDir), ...valid }, rejectedModels: rejected };
}

/** Ceny bez nadpisań z ustawień: wbudowane + opcjonalny plik użytkownika. */
export function defaultPricingTable(config: PricingConfig, homeDir: string): PricingTable {
  const filePath = config.pricingFile?.trim();
  return loadPricingTable(filePath ? expandHome(filePath, homeDir) : undefined);
}

export interface ParsedOverrides {
  valid: PricingTable;
  /** Klucze z niepoprawnym kształtem wpisu (posortowane). */
  rejected: string[];
}

export function parsePricingOverrides(raw: unknown): ParsedOverrides {
  const valid: PricingTable = {};
  const rejected: string[] = [];
  if (isPlainObject(raw)) {
    for (const [model, entry] of Object.entries(raw)) {
      if (isModelPricing(entry)) {
        valid[model] = entry;
      } else {
        rejected.push(model);
      }
    }
  }
  return { valid, rejected: rejected.sort() };
}

/**
 * Wiersz edytora cennika: brak `defaultPricing` oznacza model dodany przez użytkownika,
 * `pricing === null` — model widziany w logach, który jeszcze nie ma ceny.
 */
export interface PricingRow {
  model: string;
  pricing: ModelPricing | null;
  defaultPricing?: ModelPricing;
}

/** Modele z cenami domyślnymi (z naniesionymi nadpisaniami), modele własne, na końcu niewycenione z logów. */
export function buildPricingRows(
  defaults: PricingTable,
  overrides: PricingTable,
  unpricedModels: readonly string[] = [],
): PricingRow[] {
  const rows: PricingRow[] = Object.entries(defaults).map(([model, defaultPricing]) => ({
    model,
    pricing: overrides[model] ?? defaultPricing,
    defaultPricing,
  }));
  for (const [model, pricing] of Object.entries(overrides)) {
    if (!(model in defaults)) {
      rows.push({ model, pricing });
    }
  }
  for (const model of unpricedModels) {
    if (!(model in defaults) && !(model in overrides)) {
      rows.push({ model, pricing: null });
    }
  }
  return rows;
}

export type EditedOverridesResult =
  | { ok: true; overrides: PricingTable }
  | { ok: false; error: string };

/**
 * Wiersze odesłane przez edytor (niezaufane) → nadpisania do zapisu: nowe modele
 * i te, których cena różni się od domyślnej.
 */
export function overridesFromEditedRows(rows: unknown, defaults: PricingTable): EditedOverridesResult {
  if (!Array.isArray(rows)) {
    return { ok: false, error: "Niepoprawne dane z edytora." };
  }
  const overrides: PricingTable = {};
  const seen = new Set<string>();
  for (const row of rows) {
    const model = isPlainObject(row) && typeof row.model === "string" ? row.model.trim() : "";
    if (model === "") {
      return { ok: false, error: "Każdy model musi mieć nazwę." };
    }
    if (seen.has(model)) {
      return { ok: false, error: `Model „${model}” występuje więcej niż raz.` };
    }
    seen.add(model);
    const pricing = isPlainObject(row) ? row.pricing : undefined;
    if (!isModelPricing(pricing)) {
      return { ok: false, error: `Ceny modelu „${model}” muszą być liczbami ≥ 0.` };
    }
    const entry = pickPrices(pricing);
    const defaultPricing = defaults[model];
    if (!defaultPricing || PRICE_FIELDS.some((field) => defaultPricing[field] !== entry[field])) {
      overrides[model] = entry;
    }
  }
  return { ok: true, overrides };
}

function pickPrices(pricing: ModelPricing): ModelPricing {
  return {
    inputPer1M: pricing.inputPer1M,
    outputPer1M: pricing.outputPer1M,
    cacheReadPer1M: pricing.cacheReadPer1M,
    cacheWritePer1M: pricing.cacheWritePer1M,
  };
}

function expandHome(path: string, homeDir: string): string {
  return path === "~" || path.startsWith("~/") ? join(homeDir, path.slice(1)) : path;
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function isModelPricing(value: unknown): value is ModelPricing {
  return (
    isPlainObject(value) &&
    PRICE_FIELDS.every((field) => {
      const price = value[field];
      return typeof price === "number" && Number.isFinite(price) && price >= 0;
    })
  );
}

File: packages/intellij-plugin/engine/src/main/kotlin/dev/bredson/tokentracker/engine/PricingConfig.kt
package dev.bredson.tokentracker.engine

import java.nio.file.Path
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.doubleOrNull

data class ResolvedPricing(val table: PricingTable, val rejectedModels: List<String>)

data class ParsedOverrides(val valid: PricingTable, val rejected: List<String>)

private val prettyJson = Json { prettyPrint = true }

/** Nadpisania z ustawień: poprawne wpisy w kolejności z JSON-a, błędne po nazwie (posortowane). */
fun parsePricingOverrides(overridesJson: String?): ParsedOverrides {
    val valid = LinkedHashMap<String, ModelPricing>()
    val rejected = ArrayList<String>()
    parseObjectOrNull(overridesJson)?.forEach { (model, entry) ->
        val pricing = (entry as? JsonObject)?.toModelPricing()
        if (pricing != null) valid[model] = pricing else rejected += model
    }
    return ParsedOverrides(valid, rejected.sorted())
}

fun serializePricingOverrides(overrides: PricingTable): String =
    if (overrides.isEmpty()) "" else prettyJson.encodeToString(overrides)

/** Wpisy z edytora, które trzeba zapisać: nowe modele i te, których cena różni się od domyślnej. */
fun pricingOverridesFrom(edited: PricingTable, defaults: PricingTable): PricingTable =
    edited.filter { (model, pricing) -> defaults[model] != pricing }

/**
 * Wiersz edytora cennika: `defaultPricing == null` — model dodany przez użytkownika,
 * `pricing == null` — model widziany w logach, który jeszcze nie ma ceny.
 */
data class PricingRow(val model: String, val pricing: ModelPricing?, val defaultPricing: ModelPricing?)

/** Modele z cenami domyślnymi (z naniesionymi nadpisaniami), modele własne, na końcu niewycenione z logów. */
fun buildPricingRows(
    defaults: PricingTable,
    overrides: PricingTable,
    unpricedModels: List<String> = emptyList(),
): List<PricingRow> =
    defaults.map { (model, pricing) -> PricingRow(model, overrides[model] ?: pricing, pricing) } +
        overrides.filterKeys { it !in defaults }.map { (model, pricing) -> PricingRow(model, pricing, null) } +
        unpricedModels.filter { it !in defaults && it !in overrides }.map { PricingRow(it, null, null) }

// `dostawca/model`, opcjonalnie z sufiksem okna kontekstu — tak wypisuje je np. `/models` bramki API.
private val MODEL_ID_PATTERN = Regex("""[A-Za-z0-9][\w-]*/[A-Za-z0-9][\w.-]*(?:\[1m])?""")

/** `added` — nowe ID (znormalizowane jak w silniku) w kolejności z tekstu; `skipped` — ile już było w tabeli. */
data class ImportedModels(val added: List<String>, val skipped: Int)

/**
 * Wklejona lista modeli (dowolny tekst, np. wynik `/models`) → modele do dopisania
 * jako „brak ceny”. Porównuje po znormalizowanym ID, więc `anthropic/claude-x[1m]`
 * trafia na istniejący wiersz `claude-x`.
 */
fun modelsToImport(text: String, existingModels: Collection<String>): ImportedModels {
    val known = existingModels.map { normalizeModelId(it.trim()) }.toSet()
    val found = MODEL_ID_PATTERN.findAll(text).map { normalizeModelId(it.value.trimEnd('.')) }.distinct().toList()
    val added = found.filter { it !in known }
    return ImportedModels(added, skipped = found.size - added.size)
}

/**
 * Odpowiednik `pricingConfig.ts`: tabela wbudowana + opcjonalny plik + nadpisania
 * wpisane wprost (JSON z ustawień, niezaufane). Wpis nadpisania musi mieć cztery
 * skończone, nieujemne liczby — inaczej trafia do `rejectedModels`.
 */
fun resolvePricingTable(pricingFile: String?, overridesJson: String?, homeDir: Path): ResolvedPricing {
    val overrides = parsePricingOverrides(overridesJson)
    return ResolvedPricing(defaultPricingTable(pricingFile, homeDir) + overrides.valid, overrides.rejected)
}

/** Ceny bez nadpisań z ustawień: wbudowane + opcjonalny plik użytkownika. */
fun defaultPricingTable(pricingFile: String?, homeDir: Path): PricingTable =
    loadPricingTable(pricingFile?.trim()?.takeIf { it.isNotEmpty() }?.let { expandHome(it, homeDir) })

private fun expandHome(path: String, homeDir: Path): Path = when {
    path == "~" -> homeDir
    path.startsWith("~/") -> homeDir.resolve(path.substring(2))
    else -> Path.of(path)
}

private fun parseObjectOrNull(text: String?): JsonObject? {
    if (text.isNullOrBlank()) return null
    return try {
        Json.parseToJsonElement(text) as? JsonObject
    } catch (_: Exception) {
        null
    }
}

private fun JsonObject.price(key: String): Double? =
    (this[key] as? JsonPrimitive)?.takeUnless { it.isString }?.doubleOrNull?.takeIf { it.isFinite() && it >= 0 }

private fun JsonObject.toModelPricing(): ModelPricing? {
    val input = price("inputPer1M") ?: return null
    val output = price("outputPer1M") ?: return null
    val cacheRead = price("cacheReadPer1M") ?: return null
    val cacheWrite = price("cacheWritePer1M") ?: return null
    return ModelPricing(input, output, cacheRead, cacheWrite)
}
